libraries/AP_Baro/AP_Baro.cpp

```diff
@@ -430,3 +465,56 @@ bool AP_Baro::all_healthy(void) const
      }
      return _num_sensors > 0;
 }
+
+
+// update altitude_target
+// alt_target: cm above home
+void AP_Baro::update_alt_target(float alt_target)
+{
+    if (_adj_step < 0.1f) return; // disabled
+
+    if (fabsf(alt_target - _last_alt_target) > 0.02f*_adj_step) {
+        // alt target changed significantly (>2 steps), trigger GPS filter reset
+        _adj_sample_count = 0;
+        _last_alt_target = alt_target;
+    }
+    else {
+        // alt target is constant
+        if (_adj_sample_count >= _adj_delay*GPS_LOOP_FREQ) {
+            // enough GPS samples, start adjusting, with a max rate of one inc per second (0.1 m/s)
+            float difference = _gps_alt_over_home.get() - get_altitude();
+
+            // check difference with hysteresis of 3 steps
+            if (fabsf(difference) > 0.03f*_adj_step) {
+                // if we see a 1 hz pulsing as a result, we should do 1/10 steps in update() instead
+                _alt_offset.set(_alt_offset.get() + copysignf(_adj_step*0.01f, difference));
+            }
+        }
+    }
+}
+
+
+// store and buffer a single GPS alt reading
+// gps_alt: GPS altitude in cm (AMSL)
+void AP_Baro::update_gps_alt(float gps_alt)
+{
+    // re-initialize filter after alt change
+    if (_adj_sample_count == 0) {
+        _gps_alt_over_home.reset(gps_alt - _home_alt);
+    }
+    else {
+        _gps_alt_over_home.apply(gps_alt - _home_alt, GPS_DT);
+    }
+
```

Comment:
@magicrub During one test flight I had a home point, that was not on the ground (discussed here: https://groups.google.com/forum/#!topic/drones-discuss/FzWhvnpW0cI ). In that case,  (gps_alt-home_alt) is obviously not the altitude over takeoff. Is there any need to handle that case here? 

Or better fix the problem that HOME is not on the ground sometimes?